### PR TITLE
Implements requirements command as per #4959

### PR DESCRIPTION
**Pull request by ImreC**
This implements a feature according to https://github.com/pypa/pipenv/issues/4959.

### CODE REVIEW

**Comment by matteius** [comment]
@ImreC Could you merge latest main branch into your branch so the CI should pass again?

File: pipenv/cli/command.py [hunk 1]
```diff
@@ -709,5 +709,24 @@ def verify(state):
     sys.exit(0)
 
 
+@cli.command(
+    short_help="Generate a requirements.txt from Pipfile.lock.",
+    context_settings=CONTEXT_SETTINGS,
+)
+@option("--dev", is_flag=True, default=False, help="Also add development requirements.")
+@pass_state
+def reqs(state, dev=False):
```
Comment (matteius) [inline pipenv/cli/command.py:718]: Would you be open to calling this command `requirements` for clarity?

File: docs/advanced.rst [hunk 2]
```diff
@@ -199,19 +199,45 @@ development dependencies::
...
-Finally, if you wish to generate a requirements file with only the
+If you wish to generate a requirements file with only the
 development requirements you can do that too, using the ``--dev-only``
 flag::
 
     $ pipenv lock -r --dev-only
     py==1.4.34
     pytest==3.2.3
 
+Often, you would want to generate a requirements file based on your current
```
Comment (matteius) [inline docs/advanced.rst:210]: Perhaps call this out as `Sometimes` instead of `Often` -- my concern here is that I've been realizing many folks are not taking full advantage of the dependency lock hash checking by just generating a requirements.txt and installing from that instead.   It would be good to make the documentation more clear about the benefits of using `pipenv install --deploy` or `pipenv sync` instead of generating requirements files because of the benefit of verification that the package version matches the same signature from when you locked/pinned to it.

File: tests/integration/test_reqs.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            {packages[0]}= "=={packages[1]}"
+            [dev-packages]
+            {dev_packages[0]}= "=={dev_packages[1]}"
+            """.strip()
+            f.write(contents)
+        p.pipenv('lock')
+        c = p.pipenv('reqs')
+        assert c.returncode == 0
+
+        assert '-i https://pypi.org/simple' in c.stdout
```
Comment (matteius) [inline tests/integration/test_reqs.py:55]: Users should be aware that generating a requirements.txt with pinned indexes will result in a pip command that will not fully respect the restricted indexes.   This was something that came up recently where we now have patched the pip internals within pipenv to ensure restricted indexes are obeyed, but this is another case where the end user only gets that benefit if they use pipenv to sync or install --deploy their dependencies, because pip itself will prefer a package in pypi and just use extra-indexes to search for things not in pypi, basically (it is probably a little more complicated than that but in a nutshell).

File: docs/advanced.rst (hunk 2 above)
Comment (ImreC) [reply docs/advanced.rst:210]: I completely agree. This should only be used for the use case of interfacing with tooling that only supports requirements.txt.

File: pipenv/cli/command.py (hunk 1 above)
Comment (ImreC) [reply pipenv/cli/command.py:718]: Most definitely. I wanted to call it that, but requirements is a long word to type. Will update

**Comment by ImreC** [comment]
@matteius is the news file correct? I have never worked with this before

**Comment by matteius** [comment]
> @matteius is the news file correct? I have never worked with this before

It should be name `4959.feature.rst`  is the main thing, and anything that is code should be in double backticks.  I have a linting PR out that will add pre-commit hooks to the project and so at that time it will be more obvious if you are naming the news fragments wrong among other things.   I think the content itself makes sense, just remember to update the command name there as well.

File: pipenv/cli/command.py (hunk 1 above)
Comment (matteius) [reply pipenv/cli/command.py:718]: Thanks @ImreC for doing that, it will be more clear to non-English speakers I think why it exists.

File: tests/integration/test_reqs.py (hunk 3 above)
Comment (ImreC) [reply tests/integration/test_reqs.py]: I also think that this should not be the main route, but if I don't add the indexes and you use multiple, the requirements file will fail completely. Anything required on my part currently?

Comment (matteius) [reply tests/integration/test_reqs.py]: I think you're good, I just was calling that out because the test reminded me of that fact.  I kicked off the test CI.

File: pipenv/cli/command.py [hunk 4]
```diff
@@ -709,5 +709,24 @@ def verify(state):
     sys.exit(0)
 
 
+@cli.command(
+    short_help="Generate a requirements.txt from Pipfile.lock.",
+    context_settings=CONTEXT_SETTINGS,
+)
+@option("--dev", is_flag=True, default=False, help="Also add development requirements.")
```
Comment (oz123) [inline pipenv/cli/command.py:716]: Can you please add a flag for including the hashes?
Also, some people like keeping `requirements.txt` and `requirements-dev.txt`. I think this would be nice to have ``--dev-only`` too.

**Review by oz123 — CHANGES_REQUESTED** [review]

File: pipenv/cli/command.py (hunk 4 above)
Comment (matteius) [reply pipenv/cli/command.py:716]: Yeah we have another ticket that was just opened about getting hashes into the requirements.txt  -- https://github.com/pypa/pipenv/issues/5025

File: pipenv/cli/command.py [hunk 5]
```diff
@@ -709,5 +709,24 @@ def verify(state):
...
+@pass_state
+def requirements(state, dev=False):
+    lockfile = state.project.lockfile_content
+    for i, package_index in enumerate(lockfile['_meta']['sources']):
+        prefix = '-i' if i == 0 else '--extra-index-url'
+        echo(crayons.normal(' '.join([prefix, package_index['url']])))
```
Comment (rittneje) [inline pipenv/cli/command.py:722]: This should output the index verbatim as specified in the Pipfile, _without_ expanding variables. It is fairly common to specify credentials for a private PyPI this way, and writing those into the resulting requirements.txt file is counter-productive.

I'm not sure whether your existing code is already doing this or not, just wanted to mention this need.

File: pipenv/cli/command.py (hunk 4 above)
Comment (ImreC) [reply pipenv/cli/command.py:716]: I will add this today

File: pipenv/cli/command.py (hunk 5 above)
Comment (ImreC) [reply pipenv/cli/command.py:724]: @rittneje interesting point. I did a couple of tests and for me the $ENV_VAR gets substituted in the output and ${ENV_VAR} does not. I am currently just echoing the value from Pipfile.lock to stdout. Do you have any suggestions how to prevent this substitution in a smart way? I am still thinking about it

File: pipenv/cli/command.py [hunk 6]
```diff
@@ -714,17 +714,22 @@ def verify(state):
...
-        echo(crayons.normal(f"{req_name}{value['version']}"))
-    if dev:
+    if not dev_only:
+        for req_name, value in lockfile['default'].items():
+            hashes = [f' \\\n    --hash={h}' for h in value['hashes'] if hash]
+            echo(crayons.normal(''.join([req_name, value['version'], *hashes])))
+    if dev or dev_only:
         for req_name, value in lockfile['develop'].items():
-            echo(crayons.normal(f"{req_name}{value['version']}"))
+            hashes = [f' \\\n    --hash={h}' for h in value['hashes'] if hash]
```
Comment (oz123) [inline pipenv/cli/command.py:731]: I might be wrong here, but it seems to me you are checking if hash for every `value['hashes']`.
Can you rewrite this as:
```
if hash:
    hashes = [f' \\\n    --hash={h}' for h in value['hashes'] if hash]
else:
    hashes = []
```

**Review by oz123 — CHANGES_REQUESTED** [review]

**Comment by oz123** [comment]
@ImreC thank you for the latest changes! really good to see this, just a small change still, and I think this can be merged.

File: pipenv/cli/command.py (hunk 6 above)
Comment (ImreC) [reply pipenv/cli/command.py]: Fixed

**Comment by ImreC** [comment]
For now I have changed the test back so it passes. All checks that I have performed indicates that this works as expected (the literal index URL including env var keys gets outputted with no expansion)

**Comment by matteius** [comment]
@ImreC I suspect for that test that shows the environment variables were not expanded, I think they are expanded.  It has been in known issue reports, here are just some:
https://github.com/pypa/pipenv/issues/3751
https://github.com/pypa/pipenv/issues/3138

**Comment by matteius** [comment]
@ImreC I think it does explain your test failing and unless @rittneje has some inputs on this one, we may have to defer the environment variable in the pip indexes as a separate issue/enhancement.

For your second question, I am of the view that we should try to figure out how to support pinned indexes such as `requests = {index = "pypi",version = "*"}).` -- It is interesting though, because you are generating pip commands and I have patched pip in pipenv to be able to pin indexes correctly, meaning that the pip command is likely going to do it funky, potentially even wrong, by searching pypi and other indexes URLs for something that is index restricted.

**Comment by matteius** [comment]
Ok so `requests = {index = "pypi",version = "*"})` isn't the best example because pypi is the default index and that basically gets pinned to pypi when when you leave off the index such as ` requests = {version = "*"})` ... a better example is torch:

```
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[[source]]
url = "https://download.pytorch.org/whl/cu113/"
verify_ssl = true
name = "downloadpytorch"

[packages]
torch = {version = "==1.11.0+cu113", index = "downloadpytorch"}

[dev-packages]

[requires]
python_version = "3.10"
```

**Comment by ImreC** [comment]
I would definitely leave it in for now as the behavior is different than `pipenv requirements`, but it might make more sense to at some point take it out. Maybe `pipenv lock` does too much with the options currently in it. Conceptually it is maybe cleaner to use `pipenv lock` for creating a lockfile and doing that well, and then using `pipenv requirements` to generate requirements.txt from that, instead of trying to do everything from a single command with flags. I have not looked into the implementation, but happy to have a look at some point to see what it currently does.

**Comment by matteius** [comment]
@ImreC Looks like your build is failing on the new linting checks I added recently.  Can you run `pre-commit install` on your branch and cleanup any errors generated during: `pre-commit run --all-files --verbose`.   `pip install pre-commit` if you don't have it already.

**Comment by ImreC** [comment]
@matteius should be fixed

**Review by matteius — APPROVED** [review]

**Comment by oz123** [comment]
@ImreC you might want to change your author email. It seems currently wrong. If you are fine with the current email, I can merge it like this.

**Comment by ImreC** [comment]
@oz123 I changed my email address mid-PR so I think it is fine. Just for my understanding. Where is this visible?

**Comment by ImreC** [comment]
> @matteius suppression expansion of variables is already merged into pip and I believe also into pipenv. So this should not happen here. Also, as @ImreC already mentioned index url are not part of `requirements.txt`. I believe pip can handle the following:
> 
> ```
> $ cat requirements.txt
> package_a==1.0.1 # from private pypi
> package_b==2.0.2 # from another pypi
> requests  # from pypi
> ```
> 
> can be resolved with:
> 
> ```
> pip install --extra-index-url https://private_a.org l --extra-index-url https://private_b.org
> ```
> 
> See also: https://stackoverflow.com/a/41926975

@oz123 I think this case will work fine. The only problem that I can see is where you have packages with clashing names, f.e. you have a private package index with a package called `requests` in it. This will definitely not work with requirements.txt, even though I think pipenv offers the option to specifically define that this package needs to come from the private index and not pypi. Or maybe similarly named packages exist in multiple locations. We could solve this by defining multiple requirements.txt outputs, but it seems like a really exotic case to me. In case you control the private index a more obvious thing to do would be to not name your package `requests`, or add prefixes or something.

**Comment by matteius** [comment]
>  The only problem that I can see is where you have packages with clashing names, f.e. you have a private package index with a package called requests in it. This will definitely not work with requirements.txt, even though I think pipenv offers the option to specifically define that this package needs to come from the private index and not pypi. 

@ImreC Actually this shouldn't be a big issue anymore, please see:   https://github.com/pypa/pipenv/pull/5029

TLDR:  Basically all packages have to be pinned now to an index, and by default the first index in the file is the primary index that unspecified packages utilize.

**Review by oz123 — APPROVED** [review]
